Pyrado/tests/algorithms/test_stopping_criteria.py

```diff
@@ -198,60 +213,141 @@ def test_criterion_rollout_based_no_sampler():
-    sampler = RolloutSavingWrapper(MockSampler([rollout_a, rollout_b, rollout_c]))
+def test_criterion_rollout_based_convergence_history_filling():
+    rollouts = [SimpleNamespace(undiscounted_return=(lambda k: lambda: k)(n)) for n in range(10)]
+    mock_sampler = MockSampler()
+    sampler = RolloutSavingWrapper(mock_sampler)
+    algo = SimpleNamespace(sampler=sampler)
+    criterion = ConvergenceStoppingCriterion()
+    for i, rollout in enumerate(rollouts):
+        mock_sampler.step_sequences = [rollout]
+        sampler.sample()
+        criterion.is_met(algo)
+        assert criterion._return_statistic_history == np.arange(i + 1).tolist()
+
+
+def test_criterion_rollout_based_convergence_regress_constant_zero():
+    criterion = ConvergenceStoppingCriterion()
+    criterion._return_statistic_history = np.zeros(10).tolist()
+    assert np.isclose(criterion._compute_convergence_probability(), 1.0)
+
+
+def test_criterion_rollout_based_convergence_regress_constant_one():
+    criterion = ConvergenceStoppingCriterion()
+    criterion._return_statistic_history = np.ones(10).tolist()
+    assert np.isclose(criterion._compute_convergence_probability(), 1.0)
+
+
+def test_criterion_rollout_based_convergence_regress_not_constant():
+    criterion = ConvergenceStoppingCriterion()
+    criterion._return_statistic_history = np.arange(10).tolist()
+    assert np.isclose(criterion._compute_convergence_probability(), 0.0)
+
+
+def test_criterion_rollout_based_convergence_regress_random():
+    criterion = ConvergenceStoppingCriterion()
+    criterion._return_statistic_history = np.random.default_rng(seed=5).normal(loc=0.0, scale=0.001, size=10000)
+    assert criterion._compute_convergence_probability() > 0.9
+
+
+def test_criterion_rollout_based_convergence_lower():
+    rollout = SimpleNamespace(undiscounted_return=lambda: 0)
+    sampler = RolloutSavingWrapper(MockSampler([rollout]))
     sampler.sample()
     algo = SimpleNamespace(sampler=sampler)
-    criterion = MinReturnStoppingCriterion(min_return=2)
+    criterion = ConvergenceStoppingCriterion()
+    criterion._compute_convergence_probability = mock.MagicMock()
+    criterion._compute_convergence_probability.return_value = 0.0
     assert not criterion.is_met(algo)
 
 
-# noinspection PyTypeChecker
-def test_criterion_rollout_based_min_return_use_last():
-    rollout_a = SimpleNamespace(undiscounted_return=lambda: 3)
-    rollout_b = SimpleNamespace(undiscounted_return=lambda: 2)
-    rollout_c = SimpleNamespace(undiscounted_return=lambda: 1)
-    sampler = RolloutSavingWrapper(MockSampler([rollout_a, rollout_b, rollout_c]))
+def test_criterion_rollout_based_convergence_higher():
+    rollout = SimpleNamespace(undiscounted_return=lambda: 0)
+    sampler = RolloutSavingWrapper(MockSampler([rollout]))
+    sampler.sample()
+    algo = SimpleNamespace(sampler=sampler)
+    criterion = ConvergenceStoppingCriterion()
+    criterion._compute_convergence_probability = mock.MagicMock()
```

Comment:
Ah, why do you use MagicMock here?

Reply:
To not having to find a set of rollouts that produces this exact p-value, i.e. to specify the return value of the `_compute_convergence_probability` function.

Reply:
But why not `lambda: 0.0`?